src/store.rs

```diff
@@ -1,110 +1,251 @@
 //! Store traits.
 use crate::block::Block;
 use crate::cid::Cid;
-use crate::codec::{Codec, Decode};
+use crate::codec::{Codec, Decode, Encode};
 use crate::error::Result;
 use crate::ipld::Ipld;
 use crate::multihash::MultihashDigest;
 use crate::path::DagPath;
-use core::future::Future;
-use core::pin::Pin;
-use std::path::Path;
+use async_trait::async_trait;
+use std::collections::{HashSet, VecDeque};
 
-/// Result type of store methods.
-pub type StoreResult<'a, T> = Pin<Box<dyn Future<Output = Result<T>> + Send + 'a>>;
-
-/// Implementable by ipld storage providers.
-pub trait ReadonlyStore: Clone + Send + Sync {
+/// The store parameters.
+pub trait StoreParams: Clone + Send + Sync {
     /// The multihash type of the store.
-    type Multihash: MultihashDigest;
+    type Hashes: MultihashDigest;
     /// The codec type of the store.
-    type Codec: Codec;
+    type Codecs: Codec;
     /// The maximum block size supported by the store.
     const MAX_BLOCK_SIZE: usize;
+}
+
+/// Default store parameters.
+#[derive(Clone)]
+pub struct DefaultStoreParams;
+
+impl StoreParams for DefaultStoreParams {
+    const MAX_BLOCK_SIZE: usize = crate::MAX_BLOCK_SIZE;
```

Comment:
As I'm still working on my own Block API I don't want to block this version of the of a Block API. What I care about that on the Block level there is by default no size limit. What I can live with is if the `DefaultStoreParams` set the `MAX_BLOCK_SIZE` to `usize::MAX`.

Reply:
The block is constructed for a store. Any practical store will have some limitation. Sure we can set it to usize::MAX.

Reply:
I'm not sure if binding a Block to a certain store is the right coupling. I would rather have a Block that works independent of the store. Anyway, as I said, I've a different idea of a Block API, so we'll see how all this develops in the future.

Reply:
Can you elaborate on the deficiencies in the current api?

Reply:
I currently can't I switched to concentrate on getting tiny-multihash upstreamed, so I haven't had time to work on the block API. As I don't know when I'll have time, I'm OK with merging the current state (with `usize::MAX`) and have a closer look some time later.

Reply:
I'm confused. If there is nothing wrong with the current api, why change it?

Reply:
I care that there is no block size limit (or a very high one).

I haven't had time to take a closer look at the overall API, hence I can't propose any other changes or say that it works for me.

Reply:
So basically I think the API is fine, but I'd like to take a closer look at some point. The only thing that I immediately saw was the size limit, which I don't like.